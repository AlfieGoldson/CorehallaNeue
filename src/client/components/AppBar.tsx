// Library imports
import React, { FC, useState } from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { motion, AnimatePresence } from 'framer-motion';
import { Icon } from '@mdi/react';
import { mdiArrowLeft, mdiMagnify } from '@mdi/js';

// Custom Hooks imports
import { useScrollPosition } from '../hooks/useScrollPosition';

// Components imports
import { SearchBar } from './SearchBar';

interface NavbarProps {
    title?: string;
}

const NavbarWrapper = styled.div<{ showSearch?: boolean }>`
    display: flex;
    justify-content: space-between;
    align-items: center;
    z-index: 101;
    position: relative;
    border-bottom: 1px solid var(--bg-1);
    height: 3rem;

    & > div {
        display: flex;
        align-items: center;
    }

    ${({ showSearch }) => `
    padding: ${showSearch ? '0 1rem 0 0' : '1rem'};
        background-color: var(${showSearch ? '--text' : '--bg'});
        color: var(${showSearch ? '--bg' : '--text'});
        svg {
            fill: var(${showSearch ? '--bg' : '--text'});
        }
    `}
`;

const NavbarTitle = styled.span`
    font-size: 1rem;
    margin-left: 1rem;
`;

const NavbarTitleLogo = styled.img`
    height: 1.5rem;
`;

export const Navbar: FC<NavbarProps> = ({ title }: NavbarProps) => {
    const [showSearch, setShowSearch] = useState(false);

    return (
        <NavbarWrapper showSearch={showSearch}>
            {showSearch ? (
                <SearchBar setShowSearch={setShowSearch} />
            ) : (
                <>
                    {title ? (
                        <div>
                            <Link to="/">
                                <Icon path={mdiArrowLeft} size={1} />
                            </Link>
                            <NavbarTitle>{title}</NavbarTitle>{' '}
                        </div>
                    ) : (
                        <Link to="/">
                            <NavbarTitleLogo src="/assets/images/logo.png" alt="corehalla logo" />
                        </Link>
                    )}
                    <a
                        href="#"
                        onClick={(e) => {
                            e.preventDefault();
                            setShowSearch(true);
                        }}
                    >
                        <Icon path={mdiMagnify} size={1} />
                    </a>
                </>
            )}
        </NavbarWrapper>
    );
};

interface ITab {
    title: string;
    link: string;
    active?: boolean;
}

interface TabsProps {
    tabs: ITab[];
}

const Tab = styled.div<{ active?: boolean }>`
    white-space: nowrap;
    font-size: 0.75rem;
    text-transform: uppercase;
    a {
        height: 100%;
        color: var(--text);
        opacity: 0.48;
        display: flex;
        align-items: center;
        padding: 0 1.5rem;
    }

    ${({ active }) =>
        active &&
        `
        a {
            opacity: 1;
            color: var(--accent);
            border-bottom:1px solid var(--accent);
        }
    `}
`;

const TabsContainerWrapper = styled.div`
    display: flex;
    overflow-x: auto;
    border-bottom: 1px solid var(--bg-1);
    background-color: var(--bg);
    height: 3rem;
`;

export const TabsContainer: FC<TabsProps> = ({ tabs }: TabsProps) => {
    return (
        <TabsContainerWrapper>
            {tabs.map(({ title, link, active }, i) => (
                <Tab key={i} active={active}>
                    <Link to={link}>{title}</Link>
                </Tab>
            ))}
        </TabsContainerWrapper>
    );
};

interface IChip {
    title: string;
    action?: () => void;
    link?: string;
    active?: boolean;
}

interface ChipsContainerProps {
    chips: IChip[];
}

const Chip = styled.div<{ active?: boolean }>`
    white-space: nowrap;
    margin: 0 0.5rem;
    border-radius: 2rem;
    border: 1px solid var(--bg-3);
    background-color: var(--bg-1);
    a {
        padding: 0.25rem 1rem;
        display: block;
        color: var(--text);
    }

    ${({ active }) =>
        active &&
        `
        background-color: var(--accent);
        border-color: var(--accent);
        opacity: 1;
        a {
            color: var(--bg);
        }
    `}
`;

const ChipsContainerWrapper = styled.div`
    font-size: 1rem;
    display: flex;
    align-items: center;
    overflow-x: auto;
    border-bottom: 1px solid var(--bg-1);
    background-color: var(--bg);
    height: 3rem;
`;

export const ChipsContainer: FC<ChipsContainerProps> = ({ chips }: ChipsContainerProps) => {
    return (
        <ChipsContainerWrapper>
            {chips.map(({ title, action, active, link }, i) => (
                <Chip key={i} active={active}>
                    <Link
                        to={link || '#'}
                        onClick={
                            action
                                ? (e) => {
                                      e.preventDefault();
                                      action();
                                  }
                                : undefined
                        }
                    >
                        {title}
                    </Link>
                </Chip>
            ))}
        </ChipsContainerWrapper>
    );
};

interface Props {
    title?: string;
    tabs?: ITab[];
    chips?: IChip[];
}

const AppBarWrapper = styled.div`
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    z-index: 100;
`;

export const AppBar: FC<Props> = ({ title, tabs, chips }: Props) => {
    const [hideOnScroll, setHideOnScroll] = useState(false);

    useScrollPosition(
        ({ currScrollPos, prevScrollPos }) => {
            const isShow = currScrollPos > prevScrollPos;
            if (isShow !== hideOnScroll) setHideOnScroll(isShow);
        },
        [hideOnScroll],
    );

    return (
        <AppBarWrapper>
            <Navbar title={title} />
            <AnimatePresence>
                {!hideOnScroll && (
                    <motion.div
                        animate={{ opacity: 1, y: 0 }}
                        exit={{ opacity: 0, y: -80 }}
                        transition={{ ease: 'linear', duration: 0.125 }}
                    >
                        {tabs && <TabsContainer tabs={tabs} />}
                        {chips && <ChipsContainer chips={chips} />}
                    </motion.div>
                )}
            </AnimatePresence>
        </AppBarWrapper>
    );
};
