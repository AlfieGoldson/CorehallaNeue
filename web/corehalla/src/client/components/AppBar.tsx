// Library imports
import React, { FC, useState } from 'react';
import { useHistory } from 'react-router-dom';
import Link from 'next/link';
import styled from 'styled-components';
import { motion, AnimatePresence, Variants } from 'framer-motion';
import { Icon } from '@mdi/react';
import { mdiArrowLeft, mdiMagnify } from '@mdi/js';

// Custom Hooks imports
import { useScrollPosition } from '../hooks/useScrollPosition';

// Components imports
import { SearchBar } from './SearchBar';

interface NavbarProps {
    title?: string;
}

const NavbarWrapper = styled.div<{ showSearch?: boolean }>`
    display: flex;
    justify-content: space-between;
    align-items: center;
    z-index: 101;
    position: relative;
    border-bottom: 1px solid var(--bg-1);
    height: 2rem;

    & > div {
        display: flex;
        align-items: center;
    }

    ${({ showSearch }) => `
    padding: ${showSearch ? '0 1rem 0 0' : '1rem'};
        background-color: var(${showSearch ? '--text' : '--bg'});
        color: var(${showSearch ? '--bg' : '--text'});
    `}
`;

const NavbarTitle = styled.span`
    font-size: 1rem;
    margin-left: 1rem;
`;

const NavbarTitleLogo = styled.img`
    height: 1.5rem;
`;

export const Navbar: FC<NavbarProps> = ({ title }: NavbarProps) => {
    const [showSearch, setShowSearch] = useState(false);
    const history = useHistory();

    return (
        <NavbarWrapper showSearch={showSearch}>
            {showSearch ? (
                <SearchBar setShowSearch={setShowSearch} />
            ) : (
                <>
                    {title ? (
                        <div>
                            {/* TODO: Button */}
                            <button
                                onClick={(e) => {
                                    e.preventDefault();
                                    history.goBack();
                                }}
                            >
                                <Icon path={mdiArrowLeft} size={1} color="var(--text)" />
                            </button>
                            <NavbarTitle>{title}</NavbarTitle>
                        </div>
                    ) : (
                        <Link href="/">
                            <NavbarTitleLogo src="/assets/images/logo.png" alt="corehalla logo" />
                        </Link>
                    )}
                    <a
                        href="#"
                        onClick={(e) => {
                            e.preventDefault();
                            setShowSearch(true);
                        }}
                    >
                        <Icon path={mdiMagnify} size={1} color="var(--text)" />
                    </a>
                </>
            )}
        </NavbarWrapper>
    );
};

interface ITab<T extends string> {
    displayName: T;
    link: string;
    active?: boolean;
}

interface TabsProps<T extends string> {
    tabs: ITab<T>[];
}

const Tab = styled.div<{ active?: boolean }>`
    white-space: nowrap;
    font-size: 0.75rem;
    text-transform: uppercase;
    a {
        height: 100%;
        color: var(--text-2);
        display: flex;
        align-items: center;
        padding: 0 1.5rem;
    }

    ${({ active }) =>
        active &&
        `
        a {
            color: var(--accent);
            border-bottom:1px solid var(--accent);
        }
    `}
`;

const TabsContainerWrapper = styled.div`
    display: flex;
    overflow-x: auto;
    border-bottom: 1px solid var(--bg-1);
    background-color: var(--bg);
    height: 3rem;
`;

export function TabsContainer<T extends string>({ tabs }: TabsProps<T>): React.ReactElement<TabsProps<T>> {
    return (
        <TabsContainerWrapper>
            {tabs.map(({ displayName, link, active }, i) => (
                <Tab key={i} active={active}>
                    <Link href={link} replace>
                        {displayName}
                    </Link>
                </Tab>
            ))}
        </TabsContainerWrapper>
    );
}

interface IChip<T extends string> {
    displayName: T;
    action?: () => void;
    active?: boolean;
}

interface ChipsContainerProps<T extends string> {
    chips: IChip<T>[];
}

const Chip = styled.div<{ active?: boolean }>`
    white-space: nowrap;
    margin: 0 0.5rem;
    border-radius: 2rem;
    border: 1px solid var(--bg-2);
    background-color: var(--bg-1);
    a {
        padding: 0.25rem 1rem;
        display: block;
        color: var(--text);
    }

    ${({ active }) =>
        active &&
        `
        background-color: var(--accent);
        border-color: var(--accent);
        a {
            color: var(--bg);
        }
    `}
`;

const ChipsContainerWrapper = styled.div`
    font-size: 1rem;
    display: flex;
    align-items: center;
    overflow-x: auto;
    border-bottom: 1px solid var(--bg-1);
    background-color: var(--bg);
    height: 3rem;
`;

export function ChipsContainer<T extends string>({
    chips,
}: ChipsContainerProps<T>): React.ReactElement<ChipsContainerProps<T>> {
    return (
        <ChipsContainerWrapper>
            {chips.map(({ displayName, action, active }, i) => (
                <Chip key={i} active={active}>
                    <Link
                        to="#"
                        onClick={
                            action
                                ? (e) => {
                                      e.preventDefault();
                                      action();
                                  }
                                : null
                        }
                    >
                        {displayName}
                    </Link>
                </Chip>
            ))}
        </ChipsContainerWrapper>
    );
}

export interface AppBarProps<T extends string, U extends string> {
    title?: string;
    tabs?: ITab<T>[];
    chips?: IChip<U>[];
}

const AppBarWrapper = styled.div`
    position: sticky;
    top: 0;
    z-index: 10;
`;

const ExtrasContainer = styled.div`
    position: sticky;
    top: 2rem;
    left: 0;
    right: 0;
    height: auto;
    z-index: 9;
`;

const extrasContainerVariants: Variants = {
    in: {
        opacity: 1,
        y: 0,
        transition: { ease: 'linear', duration: 0.125 },
    },
    out: {
        opacity: 0,
        y: -80,
    },
};

export function AppBar<T extends string, U extends string>({
    title,
    tabs,
    chips,
}: AppBarProps<T, U>): React.ReactElement<AppBarProps<T, U>> {
    const [hideOnScroll, setHideOnScroll] = useState(false);

    useScrollPosition(
        ({ currScrollPos, prevScrollPos }) => {
            const isShow = currScrollPos > prevScrollPos;
            if (isShow !== hideOnScroll) setHideOnScroll(isShow);
        },
        [hideOnScroll],
    );

    return (
        <>
            <AppBarWrapper>
                <Navbar title={title} />
            </AppBarWrapper>
            <ExtrasContainer>
                <motion.div animate={hideOnScroll ? 'out' : 'in'} variants={extrasContainerVariants}>
                    {tabs && <TabsContainer tabs={tabs} />}
                    {chips && <ChipsContainer chips={chips} />}
                </motion.div>
            </ExtrasContainer>
        </>
    );
}
