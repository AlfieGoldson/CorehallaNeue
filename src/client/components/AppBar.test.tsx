import React from 'react';
import { shallow } from 'enzyme';
import { AppBar } from './AppBar';

describe('AppBar', () => {
    it('should render correctly without args', () => {
        const component = shallow(<AppBar />);
        expect(component).toMatchSnapshot();
    });
    it('should render correctly with tabs', () => {
        const component = shallow(
            <AppBar
                tabs={[
                    { title: 'title', link: 'link', active: true },
                    { title: 'title2', link: 'link2' },
                ]}
            />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should render correctly with chips', () => {
        const component = shallow(
            <AppBar
                chips={[
                    { title: 'title', link: 'link', active: true },
                    { title: 'title2', link: 'link2' },
                ]}
            />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should render correctly with tabs && chips', () => {
        const component = shallow(
            <AppBar
                chips={[
                    { title: 'title', link: 'link', active: true },
                    { title: 'title2', link: 'link2' },
                ]}
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
