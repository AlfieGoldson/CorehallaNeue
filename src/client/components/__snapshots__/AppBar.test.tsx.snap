// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar should render correctly with chips 1`] = `
<styled.div>
  <Component />
  <AnimatePresence>
    <ForwardRef(MotionComponent)
      animate={
        Object {
          "opacity": 1,
          "y": 0,
        }
      }
      exit={
        Object {
          "opacity": 0,
          "y": -80,
        }
      }
      transition={
        Object {
          "duration": 0.125,
          "ease": "linear",
        }
      }
    >
      <Component
        chips={
          Array [
            Object {
              "active": true,
              "link": "link",
              "title": "title",
            },
            Object {
              "link": "link2",
              "title": "title2",
            },
          ]
        }
      />
    </ForwardRef(MotionComponent)>
  </AnimatePresence>
</styled.div>
`;

exports[`AppBar should render correctly with tabs && chips 1`] = `
<styled.div>
  <Component />
  <AnimatePresence>
    <ForwardRef(MotionComponent)
      animate={
        Object {
          "opacity": 1,
          "y": 0,
        }
      }
      exit={
        Object {
          "opacity": 0,
          "y": -80,
        }
      }
      transition={
        Object {
          "duration": 0.125,
          "ease": "linear",
        }
      }
    >
      <Component
        chips={
          Array [
            Object {
              "active": true,
              "link": "link",
              "title": "title",
            },
            Object {
              "link": "link2",
              "title": "title2",
            },
          ]
        }
      />
    </ForwardRef(MotionComponent)>
  </AnimatePresence>
</styled.div>
`;

exports[`AppBar should render correctly with tabs 1`] = `
<styled.div>
  <Component />
  <AnimatePresence>
    <ForwardRef(MotionComponent)
      animate={
        Object {
          "opacity": 1,
          "y": 0,
        }
      }
      exit={
        Object {
          "opacity": 0,
          "y": -80,
        }
      }
      transition={
        Object {
          "duration": 0.125,
          "ease": "linear",
        }
      }
    >
      <Component
        tabs={
          Array [
            Object {
              "active": true,
              "link": "link",
              "title": "title",
            },
            Object {
              "link": "link2",
              "title": "title2",
            },
          ]
        }
      />
    </ForwardRef(MotionComponent)>
  </AnimatePresence>
</styled.div>
`;

exports[`AppBar should render correctly without args 1`] = `
<styled.div>
  <Component />
  <AnimatePresence>
    <ForwardRef(MotionComponent)
      animate={
        Object {
          "opacity": 1,
          "y": 0,
        }
      }
      exit={
        Object {
          "opacity": 0,
          "y": -80,
        }
      }
      transition={
        Object {
          "duration": 0.125,
          "ease": "linear",
        }
      }
    />
  </AnimatePresence>
</styled.div>
`;
